src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/mirrorpoint/delete/actions/DeallocateFlowMirrorPathResourcesAction.java

```diff
@@ -0,0 +1,69 @@
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ */
+
+package org.openkilda.wfm.topology.flowhs.fsm.mirrorpoint.delete.actions;
+
+import static java.lang.String.format;
+
+import org.openkilda.messaging.error.ErrorType;
+import org.openkilda.model.FlowMirrorPath;
+import org.openkilda.model.FlowMirrorPoints;
+import org.openkilda.model.PathId;
+import org.openkilda.persistence.PersistenceManager;
+import org.openkilda.persistence.repositories.FlowMirrorPathRepository;
+import org.openkilda.wfm.share.flow.resources.FlowResourcesManager;
+import org.openkilda.wfm.topology.flowhs.exception.FlowProcessingException;
+import org.openkilda.wfm.topology.flowhs.fsm.common.actions.BaseFlowPathRemovalAction;
+import org.openkilda.wfm.topology.flowhs.fsm.mirrorpoint.delete.FlowMirrorPointDeleteContext;
+import org.openkilda.wfm.topology.flowhs.fsm.mirrorpoint.delete.FlowMirrorPointDeleteFsm;
+import org.openkilda.wfm.topology.flowhs.fsm.mirrorpoint.delete.FlowMirrorPointDeleteFsm.Event;
+import org.openkilda.wfm.topology.flowhs.fsm.mirrorpoint.delete.FlowMirrorPointDeleteFsm.State;
+
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
+public class DeallocateFlowMirrorPathResourcesAction
+        extends BaseFlowPathRemovalAction<FlowMirrorPointDeleteFsm, State, Event, FlowMirrorPointDeleteContext> {
+    private final FlowResourcesManager resourcesManager;
+    private final FlowMirrorPathRepository flowMirrorPathRepository;
+
+    public DeallocateFlowMirrorPathResourcesAction(PersistenceManager persistenceManager,
+                                                   FlowResourcesManager resourcesManager) {
+        super(persistenceManager);
+        this.resourcesManager = resourcesManager;
+        this.flowMirrorPathRepository = persistenceManager.getRepositoryFactory().createFlowMirrorPathRepository();
+    }
+
+    @Override
+    protected void perform(State from, State to, Event event, FlowMirrorPointDeleteContext context,
+                           FlowMirrorPointDeleteFsm stateMachine) {
+        PathId mirrorPathId = stateMachine.getMirrorPathId();
+
+        FlowMirrorPath flowMirrorPath = flowMirrorPathRepository.findById(mirrorPathId)
+                .orElseThrow(() -> new FlowProcessingException(ErrorType.NOT_FOUND,
+                        format("Flow mirror point %s not found", mirrorPathId)));
+
+        FlowMirrorPoints flowMirrorPoints = flowMirrorPath.getFlowMirrorPoints();
+        stateMachine.setFlowPathId(flowMirrorPoints.getFlowPathId());
+        stateMachine.setMirrorSwitchId(flowMirrorPoints.getMirrorSwitchId());
+
+        resourcesManager.deallocateCookie(flowMirrorPath.getCookie().getFlowEffectiveId());
```

Comment:
Do we share cookie with target flow? If so, how we can release it on mirror point removal (while target flow still exists)?

Reply:
We do not share cookie with target flow.